### PR TITLE
Indirect - Corrections Interfaces - Add Plot Current Preview

### LINKED ISSUE #19094: Indirect - Add plot current preview
Add an option to plot the spectrum currently shown on the miniPlot preview window. Also tidy up GUI surrounding miniPlot window.
See ConvFit for example.

- [x] Bayes - #17262 
- [x] Corrections
- [x] Data Analysis

### PR DESCRIPTION
**Pull request by brandonhewer**
Currently in the Indirect Corrections interface, the ContainerSubtraction and ApplyPaalmanPings tabs do not have the functionality for plotting the current preview (mini-plot).

Added functionality for plotting current preview in the ContainerSubtraction and ApplyPaalmanPings interfaces.

**To test:**
1. Navigate to Interfaces > Indirect > Corrections.
2. To test each of the interfaces:
    1. In the ContainerSubtraction interface/tab:
        1. Enter a reduced file (*_red.nxs) or workspace (*_red) into the 'Sample' field.
        2. Once the graph in the preview (mini-plot) appears, click 'Plot Current Preview'.
        3. Ensure the plot matches the preview and then close the plot.
        4. Enter a reduced container file (*_red.nxs) into the 'Container' field.
        5. Once the graph in the preview (mini-plot) updates, click 'Plot Current Preview'.
        6. Ensure the plot matches the preview and then close the plot.
        7. Change the number in the spinner of the 'Spectrum' field.
        8. Once the graph in the preview (mini-plot) appears, click 'Plot Current Preview'.
        9. Ensure the plot matches the preview and then close the plot.
        10. Run the algorithm.
        11. Once the algorithm has completed and the preview has updated, click 'Plot Current Preview'.
        12. Ensure the plot matches the preview.
    2. In the ApplyPaalmanPings interface/tab:
        1. Enter a reduced file (*_red.nxs) or workspace (*_red) into the Input field.
        2. Once the graph in the preview (mini-plot) appears, click 'Plot Current Preview'.
        3. Ensure the plot matches the preview and then close the plot.
        4. Check the checkbox labelled 'Use Container' and enter a reduced container file/workspace into the field.
        5. Once the graph in the preview (mini-plot) updates, click 'Plot Current Preview'.
        6. Ensure the plot matches the preview and then close the plot.
        7. Change the number in the spinner of the 'Spectrum' field.
        8. Once the graph in the preview (mini-plot) updates, click 'Plot Current Preview'.
        9. Ensure the plot matches the preview and then close the plot.
        10. Run the algorithm.
        11. Once the algorithm has completed and the preview has updated, click 'Plot Current Preview'.
        12. Ensure the plot matches the preview.

Fixes #19094. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by mantid-roman — CHANGES_REQUESTED** [review]
There is a small bug:
1. Add the sample.
2. Change spectrum with the spinner.
3. Add container. Now the preview shows sample's n-th spectrum and 1st containers's spectrum.
4. 'Plot Current Preview' plots the n-th spectra form both workspaces which is not what we see in the preview.

**Comment by brandonhewer** [comment]
@mantid-roman This appears to be an issue with the preview (first spectrum of container is always loaded - regardless of what is in the 'Spectrum' field. Changing this to plot the n-th spectrum of the container when it is first loaded is a simple change, however, this brings up the issue of a container file which has a different number of spectrum than the sample file. In this scenario, a value may be entered in the 'Spectrum' field, which is out of range in the container file - of course this is invalid input and the Subtraction/Paalman Pings algorithms will deal with this situation respectively - however, it is unclear what might be expected to be seen in the preview/interface and how the issue may be conveyed to the user.

**Comment by mantid-roman** [comment]
I think the preview should try to display the n-th spectrum for both workspaces. If n is out of range for any of them then change it to a valid value and update the 'spectrum' field.

**Comment by brandonhewer** [comment]
@mantid-roman This certainly seems like a reasonable thing to do - perhaps just the minimax of the two files would work fine for this.

File: MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp [hunk 1]
```diff
@@ -481,5 +483,56 @@ void ContainerSubtraction::plotClicked() {
...
+                                         MatrixWorkspace_sptr &ws,
+                                         const QColor &curveColor) {
+
+  // Check whether the selected spectra is now out of bounds with
+  // respect to the specified workspace.
+  if (ws->getNumberHistograms() > m_spectra) {
+    m_uiForm.ppPreview->addSpectrum(curveName, ws, m_spectra, curveColor);
+  } else if (m_csSampleWS || m_csContainerWS) {
+    size_t specNo = std::min(m_csContainerWS->getNumberHistograms(),
+                             m_csSampleWS->getNumberHistograms()) -
```
Comment (mantid-roman) [inline MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp]: If one of the pointers is null shouldn't this code crash?

File: MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp [hunk 2]
```diff
@@ -481,5 +483,56 @@ void ContainerSubtraction::plotClicked() {
...
+                    1;
```
Comment (mantid-roman) [inline MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp]: Can this be moved to the previous line? I don't think clang-format will mind.

**Comment by mantid-roman** [comment]
If I have only Sample loaded and change Spectrum I get this message:

`Cannot plot with empty workspace name`.

File: MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp (hunk 1 above)
Comment (brandonhewer) [inline MantidQt/CustomInterfaces/src/Indirect/ContainerSubtraction.cpp]: Yes that's completely right, this is a mistake.

**Comment by brandonhewer** [comment]
@mantid-roman For some reason, ClangFormat appears intent on moving the '- 1' to the next line.

**Comment by mantid-roman** [comment]
Message `Cannot plot with empty workspace name` is still there.

**Comment by brandonhewer** [comment]
@mantid-roman This was occurring, as the container workspace spectrum was set to be added to the plot preview when it did not exist. The most recent update should fix this.

**Review by mantid-roman — APPROVED** [review]

**Pull request merged**